File: 78-subsets/subsets.java
class Solution {
    List<List<Integer>> ans;
    public List<List<Integer>> subsets(int[] nums) {
         ans=new ArrayList<>();
         helper(nums,0,new ArrayList<Integer>());
         return ans;
    }

    private void helper(int[] nums,int idx,ArrayList<Integer> ds){
      if(idx==nums.length){
        ans.add(new ArrayList<>(ds));
        return;
      }
      //take
      ds.add(nums[idx]);
      helper(nums,idx+1,ds);
      ds.remove(ds.size()-1);

      //not Take
      helper(nums,idx+1,ds);
    }
}

File: 78-subsets/subsets.cpp
class Solution {
public:
    vector<vector<int>>subs;
    void findSubset(int i,int n,vector<int>&nums,vector<int>&sub){
        if(i>=n){
            subs.push_back(sub);
            return;
        }
       
        
        findSubset(i+1,n,nums,sub);
        sub.push_back(nums[i]);
        findSubset(i+1,n,nums,sub);
        sub.pop_back();
    }
    
    vector<vector<int>> subsets(vector<int>& nums) {
       vector<int>sub;
       int n=nums.size();
      findSubset(0,n,nums,sub) ;
      return subs;
    }
};